src/Utils.pm

```diff
@@ -110,67 +150,294 @@ sub calcPosFromTime {
+	}
+
+	if ($field->checkPathFree($easy_solution)) {
+		return $easy_solution;
+
+	# If there are obstacles then use the client pathfinding to solve it
+	} else {
+		return get_client_solution($field, $pos, $pos_to);
+	}
+}
+
+# Currently the go-to function to get the position of a given actor on critical ocasions (eg. Attack logic)
+sub calcPosFromPathfinding {
+	my ($field, $actor, $extra_time) = @_;
+	my $speed = ($actor->{walk_speed} || 0.12);
+	my $time = time - $actor->{time_move} + $extra_time;
+
+	# If Pos and PosTo are the same return Pos
+	if ($actor->{pos}{x} == $actor->{pos_to}{x} && $actor->{pos}{y} == $actor->{pos_to}{y}) {
+		return $actor->{pos};
+	}
+
+	my $solution;
+
+	# If the actor is the character then we should have already saved the time calc and solution at Receive.pm::character_moves
+	if (UNIVERSAL::isa($actor, "Actor::You")) {
+		if ($time >= $actor->{time_move_calc}) {
+			return $actor->{pos_to};
 		}
+		$solution = $actor->{solution};
+
+	} else {
+		$solution = get_solution($field, $actor->{pos}, $actor->{pos_to});
+	}
+
+	my $steps_walked = calcStepsWalkedFromTimeAndSolution($solution, $speed, $time);
+
+	my $pos = $solution->[$steps_walked];
+
+	return $pos;
+}
+
+# Wrapper for calcTimeFromSolution so you don't need to call get_client_solution and calcTimeFromSolution when you only need the time
+# Used in Misc::meetingPosition to calculate if the target would have time to catch-up with the character when running away from it
+sub calcTimeFromPathfinding {
+	my ($field, $pos, $pos_to, $speed) = @_;
+
+	if ($pos->{x} == $pos_to->{x} && $pos->{y} == $pos_to->{y}) {
+		return 0;
 	}
 
-	%result = moveAlong($pos, $pos_to, $s);
-	return (\%result, $s);
+	my $solution = get_solution($field, $pos, $pos_to);
+
+	my $summed_time = calcTimeFromSolution($solution, $speed);
+	return $summed_time;
+}
+
+# Returns:
+# -1: No LOS
+#  0: out of range
+#  1: sucess
+#
+# Reference: hercules src\map\battle.c battle_check_range
+sub canAttack {
+	my ($field, $pos1, $pos2, $attackCanSnipe, $range, $clientSight) = @_;
+
+	my $distance = blockDistance($pos1, $pos2);
+	return 1 if ($distance < 2);
+
+	# hercules conf\map\battle\client.conf area_size
+	# Here the check is done against area_size (which is by default 14, can be higher, eg. 22 in OathRO)
+	# Openkore clientSight should be area_size+1 (by default 15)
+	return 0 if ($distance >= $clientSight);
+
+	my $client_distance = getClientDist($pos1, $pos2);
+	return 0 unless ($client_distance <= $range);
+
+	return -1 unless ($field->checkLOS($pos1, $pos2, $attackCanSnipe));
+
+	return 1;
+}
+
+# Only God and gravity developers know why this is done this way, but I tested in the client and it works 100% of the time
+#
+# Reference: hercules src\map\path.c distance_client
+sub getClientDist {
+	my ($pos1, $pos2) = @_;
+	my $xD = abs($pos1->{x} - $pos2->{x});
+	my $yD = abs($pos1->{y} - $pos2->{y});
+	my $temp_dist = sqrt(($xD*$xD) + ($yD*$yD));
+	$temp_dist -= 0.0625;
```

Comment:
can u confirm this?
https://github.com/rathena/rathena/pull/6950

Reply:
In Hercules that is the case, I would have to search some more to find if it is correct.
Source:
https://github.com/HerculesWS/Hercules/blob/69edb0cf6de382dfa128b8582617e3bc126514d4/src/map/path.c#L500C1-L500C2

Reply:
Changed to agree with rathena